docs/design/features/StringDeduplication.md

```diff
@@ -0,0 +1,226 @@
+
+```c#
+string _field;
+
+void Test()
+{
+    _field = new string('a', 1);
+
+    string a1 = _field;
+    Thread.Sleep(1000);
+    string a2 = _field;
+
+    // This assert can fail with string deduping turned on
+    Debug.Assert(ReferenceEquals(a1,a2));
+}
+```
+
+**Runtime assistance to help with these scenarios**
+
+- Static analysis
+
+For .NET 5.0, there’s a static analysis [plan](https://github.com/dotnet/runtime/issues/30740). Verifying if string content is mutated can be part of that analysis suite. It would also be beneficial to have VS do this verification and give you a warning when you are mutating string content.
+
+To start with we will provide analysis for the following –
+
+1. Using the fixed keyword to modify string content.
+
+I’m seeing that there are almost 600 places in libraries that do `fixed (char*` but hopefully most of them do not actually modify the string content. We should definitely be encouraging folks to switch to using `string.Create` like what [PR#31700](https://github.com/dotnet/runtime/pull/31700) did.
+
+2. Using lock on a string object. 
+
+- Reference equality checks on strings
+
+Since `ReferenceEquals` is performance critical API, we cannot do checks in its implementation. Checks will be done by how it's jitted - when the JIT sees that there are two object references being compared for equality, when a special diagnostics flag is set, it can jit a call to a helper method to do that and that helper method can produce diagnostics when the code is trying to compare identity of two strings.
+
+We do have some libraries that rely on `ReferenceEquals`. We need to figure out what to do about them. See discussion [here](https://github.com/dotnet/runtime/pull/31971#pullrequestreview-355531406).
+
+- Additional checks in heap verification 
+
+Heap verification will now include checks to verify that no one changes the string content after it’s hash is computed. This can be turned on when a certain level of COMPlus_HeapVerify is specified.
+
+- Stress mode
+
+Instead of waiting till the productive moment to start the next deduping cycle, we can have a stress mode where we dedup randomly to catch problems sooner, same idea as GC stress to detect GC holes sooner. 
+
+We could even artificially create duplicates in this stress mode to find places that depend on object identity.
+
+- Backup nonconcurrent dedup mode
+
+If you are running into these problematic scenarios and cannot make changes, and you can tolerate longer STW (Stop-The-World) pauses, we will provide a mode where deduping is completely done in either a full blocking GC or during the STW pause of a BGC. Obviously this would make BGC’s STW pauses much, much longer. This would basically just call the function that the dedup thread calls, but in these STW pauses. If you use a CPU profiler you should be able to see preciously how long the dedup part takes during the STW pause.
```

Comment:
The blocking mode cannot be used to workaround the compatibility problems reliably. It would work only if you can tell whether none of the threads are running the problematic code while the world is stopped which is close to impossible. All of the compat problems listed above will still be a problem even in the full blocking mode.

I doubt we would be able to recommend this as a workaround to anybody important. We cannot recommend to use something that causes crash or data corruption some of the time. Maybe as the one-shot deduplication after startup when you can know that other threads in the process are not doing anything problematic (we would need the API for that), but not as the in-flight dedup while the service is running.